walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,74 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
+  /**
+   * Get the collection of time ranges when an event can be held.
+   * @param events
+   * @param request
+   * @return Returns the collection of time ranges when the event can be held.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    Collection<TimeRange> validTimeRanges = new ArrayList<TimeRange>();
+
+    Collection<TimeRange> orderedConflicts = 
+        getOrderedAttendingEvents(events, request.getAttendees());
+    Iterator<TimeRange> conflictsIterator = orderedConflicts.iterator();
+    int prevConflictEndTime = TimeRange.START_OF_DAY;
+
+    while (conflictsIterator.hasNext()) {
+      TimeRange conflict = conflictsIterator.next();
+      int conflictStartTime = conflict.start();
+      
+      // Check for gap between prev conflict and this conflict.
+      if (conflictStartTime - prevConflictEndTime >= request.getDuration()) {
+        validTimeRanges.add(TimeRange.fromStartEnd(prevConflictEndTime, conflictStartTime, false));
+      }
+
+      // Only update conflict end time if it is greater than the existing previous end time.
+      // This is to handle nested events where previous event will have later end time.
+      if (conflict.end() > prevConflictEndTime) {
+        prevConflictEndTime = conflict.end();  
+      }
+    }
+
+    if (TimeRange.END_OF_DAY - prevConflictEndTime >= request.getDuration()) {
+      validTimeRanges.add(TimeRange.fromStartEnd(prevConflictEndTime, TimeRange.END_OF_DAY, true));  
+    }
+
+    return validTimeRanges;
+  }
+
+  /**
+   * Get a collection of the events at least one of the attendees is attending, ordered by
+   * ascending start time.
+   * @param events
+   * @param attendees
+   * @return Return a collection of the events in ascending order of start time.
+   */
+  private Collection<TimeRange> getOrderedAttendingEvents(
+      Collection<Event> events, Collection<String> attendees) {
+    ArrayList<TimeRange> timeConflicts = new ArrayList<TimeRange>();
+
+    Iterator<Event> eventsIterator = events.iterator();
+    while (eventsIterator.hasNext()) {
+      Event event = eventsIterator.next();
+      if (!Collections.disjoint(event.getAttendees(), attendees)) {
+        timeConflicts.add(event.getWhen());
+      }
+    }
+
+    Collections.sort(timeConflicts, TimeRange.ORDER_BY_START);
+    return timeConflicts;
```

Comment:
Could just return timeConflicts.iterator() here since that's all we need currently.

Reply:
Good point, although I wonder if that reduces its possible future re-usability? I tried to make it generally applicable, although it's not so complex of a function..

Reply:
Also just curious -- what is the thought process behind making a change like that? Is the idea that being as conservative as possible with the permissions of whatever we return will reduce the amount of potential bugs?

Reply:
It's not a hard and fast rule or anything. Since that's all that's needed right now and I'm not aware of why that might change, I thought I'd suggest it. Another thing to do is just call `.iterator()` immediately on line 38. 

Just generally it's mental overhead to introduce more local variables like `orderedConflicts` that aren't used anywhere else in the function. Implementation-wise I guess it might be nice for `getOrderedAttendingEvents` to take and return a Collection, but I wouldn't say an interator is somehow a less general return value.

Reply:
I'll immediately call `.iterator()` as a nice in-between. Thanks for the explanation!